File: bot/engine.py
"""Shared Overwatch anti-pick engine for the Discord bot."""
from __future__ import annotations

import json
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
_KNOWLEDGE_PATH = Path(__file__).parent / "knowledge.json"
OW = json.loads(_KNOWLEDGE_PATH.read_text(encoding="utf-8"))
HEROES = {h["key"]: h for h in OW["heroes"]}
MAPS = {m["key"]: m for m in OW["maps"]}
MATCHUPS = dict(OW.get("matchups") or {})
TAG_MATCHUPS = list(OW.get("tagMatchups") or [])
KNOWLEDGE_VERSION = ""


def apply_knowledge(data: dict, version: str = "") -> None:
    """Swap hero/map data in place so the Discord client can stay connected."""
    global KNOWLEDGE_VERSION
    if not data.get("heroes") or not data.get("maps"):
        raise ValueError("invalid knowledge.json")
    OW.clear()
    OW.update(data)
    HEROES.clear()
    HEROES.update({h["key"]: h for h in data["heroes"]})
    MAPS.clear()
    MAPS.update({m["key"]: m for m in data["maps"]})
    MATCHUPS.clear()
    MATCHUPS.update(data.get("matchups") or {})
    TAG_MATCHUPS[:] = data.get("tagMatchups") or []
    KNOWLEDGE_VERSION = version
    _KNOWLEDGE_PATH.write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding="utf-8")


def reload_from_disk() -> None:
    apply_knowledge(json.loads(_KNOWLEDGE_PATH.read_text(encoding="utf-8")))

COMP_LABEL = {
    "dive": "ダイブ",
    "poke": "ポーク",
    "brawl": "ブロウル",
    "bunker": "バンカー",
    "flying": "空中",
    "sniper": "スナイプ",
    "flank": "フランカー",
    "flex": "ミックス",
}

COMP_ANSWERS = {
    "dive": {"tank": ["orisa", "roadhog", "zarya", "dmon"], "damage": ["cassidy", "mei", "torbjorn", "reaper"], "support": ["brigitte", "ana", "moira", "wuyang"]},
    "poke": {"tank": ["winston", "dva", "doomfist", "wrecking-ball"], "damage": ["genji", "tracer", "sombra", "venture"], "support": ["lucio", "kiriko", "juno"]},
    "brawl": {"tank": ["mauga", "zarya", "junker-queen", "orisa"], "damage": ["reaper", "mei", "pharah", "bastion"], "support": ["ana", "moira", "lucio", "kiriko"]},
    "bunker": {"tank": ["wrecking-ball", "winston", "hazard", "dva"], "damage": ["sombra", "venture", "junkrat", "pharah"], "support": ["lucio", "kiriko"]},
    "flying": {"tank": ["dva", "sigma", "domina"], "damage": ["ashe", "widowmaker", "cassidy", "soldier-76", "sojourn"], "support": ["baptiste", "mizuki", "ana", "illari"]},
    "sniper": {"tank": ["winston", "dva", "wrecking-ball", "doomfist"], "damage": ["genji", "tracer", "sombra"], "support": ["lucio", "kiriko", "moira"]},
    "flank": {"tank": ["orisa", "zarya", "roadhog"], "damage": ["cassidy", "torbjorn", "mei", "symmetra"], "support": ["brigitte", "moira", "baptiste"]},
}

WEAK = {
    "flying": "ヒットスキャン不足だと空に試合を支配される",
    "dive": "CC・アンチダイブがないとバックラインが溶ける",
    "brawl": "アンチヒールと頭上火力でブロウルは崩せる",
    "poke": "接近され情報と距離を失うとポークは負ける",
    "bunker": "ハック・地中・頭上でバンカーは崩壊する",
    "sniper": "遮蔽とダイブでスナイパーを沈黙させられる",
}


def portrait_path(key: str) -> Path:
    return ROOT / "assets" / "heroes" / f"{key}.png"


def map_shot_path(key: str) -> Path:
    return ROOT / "assets" / "maps" / f"{key}.jpg"


def explicit_counter(attacker: str, target: str) -> int:
    return int((MATCHUPS.get(attacker) or {}).get(target) or 0)


def tag_counter(attacker: dict, target: dict) -> tuple[int, list]:
    a_tags = set(attacker.get("tags") or [])
    b_tags = set(target.get("tags") or [])
    best = 0
    hits = []
    for rule in TAG_MATCHUPS:
        if rule["a"] in a_tags and rule["b"] in b_tags:
            if rule["score"] >= best:
                best = rule["score"]
                hits.append(rule)
    return best, [r for r in hits if r["score"] >= 3]


def detect_composition(enemy_keys: list[str]) -> dict:
    heroes = [HEROES[k] for k in enemy_keys if k in HEROES]
    counts = {"dive": 0, "poke": 0, "brawl": 0, "bunker": 0, "flying": 0, "sniper": 0, "flank": 0}
    for h in heroes:
        t = set(h.get("tags") or [])
        if "dive" in t or "flank" in t:
            counts["dive"] += 2 if "dive" in t else 1
        if "poke" in t or "sniper" in t or "long-range" in t:
            counts["poke"] += 1
        if "brawl" in t or "melee" in t:
            counts["brawl"] += 1
        if "bunker" in t or "turret" in t:
            counts["bunker"] += 2
        if "flyer" in t:
            counts["flying"] += 2
        if "sniper" in t:
            counts["sniper"] += 2
        if "flank" in t:
            counts["flank"] += 1
    if any("flyer" in (h.get("tags") or []) for h in heroes) and any(
        "flyer-synergy" in (h.get("tags") or []) for h in heroes
    ):
        counts["flying"] += 3
    ranked = sorted(counts.items(), key=lambda kv: kv[1], reverse=True)
    primary = ranked[0][0] if ranked[0][1] > 0 else "flex"
    return {"counts": counts, "primary": primary, "heroes": heroes}


def map_affinity(hero: dict, mp: dict | None, side: str) -> tuple[int, list[str]]:
    if not mp:
        return 0, []
    traits = set(mp.get("traits") or [])
    tags = set(hero.get("tags") or [])
    score = 0
    reasons = []

    def bump(cond, pts, text):
        nonlocal score
        if cond:
            score += pts
            if pts >= 2:
                reasons.append(text)

    bump(traits.intersection({"longsight"}) and tags.intersection({"sniper", "hitscan", "long-range"}), 4, f"{mp['nameJa']}は長視線。{hero['nameJa']}の射程が刺さる")
    bump("choke" in traits and tags.intersection({"barrier", "spam", "freeze", "brawl"}), 3, f"チョークで{hero['nameJa']}のエリア制御が強い")
    bump("vertical" in traits and (tags.intersection({"flyer", "mobile"}) or hero["key"] in ("winston", "juno", "dva")), 3, "高低差で機動力が活きる")
    bump("environmental" in traits and (hero["key"] in ("lucio", "junkrat", "hazard", "pharah") or "boop" in tags), 3, "崖・井戸があり環境キルが取りやすい")
    bump("close" in traits and tags.intersection({"brawl", "melee", "tankbuster", "beam"}), 3, "近接マップで近距離キットが強い")
    bump("open" in traits and "sniper" in tags, 3, "開放マップでスナイプが通りやすい")
    bump("flashpoint" in traits and tags.intersection({"mobile", "speed", "flank"}), 3, "フラッシュポイントはローテ性能が勝つ")
    bump("clash" in traits and tags.intersection({"brawl", "melee", "barrier"}), 3, "クラッシュは超近接ブロウル")
    bump("longsight" in traits and "melee" in tags and "mobile" not in tags, -3, "長視線は近接に厳しい")
    bump("close" in traits and "sniper" in tags, -3, "近接マップでスナイパーは落ちる")
    if side == "attack" and tags.intersection({"dive", "flank", "speed"}):
        score += 1
    if side == "defend" and tags.intersection({"bunker", "barrier", "turret", "sniper"}):
        score += 1
    return score, reasons


def recommend(my_role: str, enemies: list[str], map_key: str | None, side: str = "flex") -> dict:
    my_role = my_role if my_role in ("tank", "damage", "support") else "damage"
    enemy_heroes = [HEROES[k] for k in enemies if k in HEROES]
    taken = set(enemies)
    mp = MAPS.get(map_key) if map_key else None
    comp = detect_composition(enemies)
    answers = (COMP_ANSWERS.get(comp["primary"]) or {}).get(my_role) or []
    ranked = []
    for hero in OW["heroes"]:
        if hero["role"] != my_role or hero["key"] in taken:
            continue
        score = 24
        reasons = []
        for enemy in enemy_heroes:
            direct = explicit_counter(hero["key"], enemy["key"])
            tagged, rules = tag_counter(hero, enemy)
            power = max(direct, tagged)
            threat_d = explicit_counter(enemy["key"], hero["key"])
            threat_t, _ = tag_counter(enemy, hero)
            threat = max(threat_d, threat_t * 0.7)
            score += power * 4.6 - threat * 3.4
            if power >= 4:
                reasons.append(f"{enemy['nameJa']}に対して明確なカウンター")
            elif power >= 3:
                rule = rules[0] if rules else None
                reasons.append(f"{enemy['nameJa']}：{rule['ja']}" if rule else f"{enemy['nameJa']}との相性が良い")
            if threat >= 4:
                reasons.append(f"注意：{enemy['nameJa']}から強くカウンターされる")
                score -= 2
        if hero["key"] in answers:
            score += 6
            reasons.append(f"{COMP_LABEL[comp['primary']]}構成への定番回答")
        fit, map_reasons = map_affinity(hero, mp, side)
        score += fit * 1.6
        reasons.extend(map_reasons[:2])
        if comp["primary"] == "flying" and "hitscan" in (hero.get("tags") or []):
            score += 5
            reasons.append("空中構成にはヒットスキャンが最優先")
        if comp["counts"]["flying"] >= 2 and "hitscan" in (hero.get("tags") or []):
            score += 6
            reasons.append("フォマシー／空中にはヒットスキャン")
        if comp["primary"] == "dive" and set(hero.get("tags") or []).intersection({"anti-dive", "cc", "sleeper", "anti-flank"}):
            score += 4
        if comp["primary"] == "sniper" and set(hero.get("tags") or []).intersection({"dive", "flank"}):
            score += 4
            reasons.append("スナイパー裏を取るダイブ／フランカー")
        uniq = []
        seen = set()
        for r in reasons:
            if r not in seen:
                seen.add(r)
                uniq.append(r)
        ranked.append({"hero": hero, "raw": score, "reasons": uniq[:4]})
    ranked.sort(key=lambda x: x["raw"], reverse=True)
    top_raw = ranked[0]["raw"] if ranked else 1
    floor = ranked[min(len(ranked) - 1, 12)]["raw"] if ranked else 0
    for row in ranked:
        t = (row["raw"] - floor) / max(8, top_raw - floor)
        row["score"] = round(max(12, min(99, 28 + t * 71)))
    return {
        "comp": comp,
        "comp_label": COMP_LABEL.get(comp["primary"], "ミックス"),
        "weakness": WEAK.get(comp["primary"], "構成の穴を突く"),
        "picks": ranked[:5],
        "map": mp,
    }


def parse_text(text: str) -> dict:
    lower = (text or "").lower().replace("’", "'")
    role = None
    if any(w in lower for w in ("tank", "タンク")):
        role = "tank"
    elif any(w in lower for w in ("support", "sup", "サポート", "ヒーラー")):
        role = "support"
    elif any(w in lower for w in ("dps", "damage", "ダメージ", "dps")):
        role = "damage"
    side = "flex"
    if any(w in lower for w in ("attack", "atk", "攻撃")):
        side = "attack"
    elif any(w in lower for w in ("defend", "def", "防衛", "防御")):
        side = "defend"
    map_key = None
    best = 0
    for m in OW["maps"]:
        aliases = [
            m["name"],
            m["nameJa"],
            m["key"].replace("-", " "),
            m["key"].replace("-", ""),
        ]
        for part in m["key"].split("-"):
            if len(part) >= 6:
                aliases.append(part)
        for n in aliases:
            needle = str(n).lower().replace(":", "")
            hay = lower.replace(":", "")
            if len(needle) >= 4 and needle in hay and len(needle) > best:
                map_key = m["key"]
                best = len(needle)
    hero_keys = []
    names = []
    for h in OW["heroes"]:
        names.append((h["key"], h["name"].lower()))
        names.append((h["key"], h["nameJa"].lower()))
        names.append((h["key"], h["key"].replace("-", " ")))
        names.append((h["key"], h["name"].lower().replace(":", "").replace("'", "")))
    names.sort(key=lambda x: len(x[1]), reverse=True)
    used = set()
    for key, name in names:
        if key in used or len(name) < 3:
            continue
        if name in lower:
            hero_keys.append(key)
            used.add(key)
    return {"role": role, "side": side, "map_key": map_key, "hero_keys": hero_keys}


def movement_lines(hero: dict, rec: dict, side: str) -> list[str]:
    lines = []
    mp = rec.get("map")
    coach = (mp or {}).get("coach") or {}
    if coach:
        if side == "attack":
            lines.append(coach.get("attack") or "")
        elif side == "defend":
            lines.append(coach.get("defend") or "")
        else:
            lines.append(coach.get("move") or "")
    if hero.get("play"):
        lines.append(hero["play"])
    tags = set(hero.get("tags") or [])
    primary = rec["comp"]["primary"]
    if primary == "flying" and "hitscan" in tags:
        lines.append("空の軌道は本線の真上ではなく、やや後ろの高所から切る。落ちたらすぐ次の箱へ。")
    if primary == "dive" and tags.intersection({"anti-dive", "sleeper", "cc"}):
        lines.append("自分から前に出ない。着地地点（サポの横）でCCを待つ。")
    if primary == "brawl" and "antiheal" in tags:
        lines.append("最初のアンチをタンク足下に入れて、近接の回復勝負を切る。")
    if primary == "sniper" and tags.intersection({"dive", "flank"}):
        lines.append("本線を歩かない。壁の裏から高所のスナイパーへ。")
    return [x for x in lines if x][:5]


def cd_lines(heroes: list[dict]) -> list[str]:
    out = []
    for h in heroes[:5]:
        bits = []
        for a in h.get("abilities") or []:
            cd = a.get("cd") or ""
            if cd and cd not in ("no CD", "hold", "weapon", "refresh"):
                bits.append(f"{a['nameJa']} {cd}")
            if len(bits) >= 3:
                break
        if bits:
            out.append(f"{h['nameJa']}：{' / '.join(bits)}")
    return out

File: js/engine.js
/* Anti-pick recommendation engine */
(function () {
  const OW = window.OW;
  const ROLE_ORDER = { tank: 0, damage: 1, support: 2 };

  function heroByKey(key) {
    return OW.heroes.find((h) => h.key === key);
  }

  function mapByKey(key) {
    return OW.maps.find((m) => m.key === key);
  }

  function explicitCounter(attacker, target) {
    const row = OW.matchups[attacker];
    if (!row) return 0;
    return row[target] || 0;
  }

  function tagCounter(attacker, target) {
    let best = 0;
    let hits = [];
    const aTags = new Set(attacker.tags);
    const bTags = new Set(target.tags);
    for (const rule of OW.tagMatchups) {
      if (aTags.has(rule.a) && bTags.has(rule.b)) {
        if (rule.score >= best) {
          best = rule.score;
          hits.push(rule);
        }
      }
    }
    return { score: best, rules: hits.filter((r) => r.score >= 3) };
  }

  function detectComposition(enemyKeys) {
    const heroes = enemyKeys.map(heroByKey).filter(Boolean);
    const counts = {
      dive: 0,
      poke: 0,
      brawl: 0,
      bunker: 0,
      flying: 0,
      sniper: 0,
      flank: 0,
    };
    for (const h of heroes) {
      const t = new Set(h.tags);
      if (t.has("dive") || t.has("flank")) counts.dive += t.has("dive") ? 2 : 1;
      if (t.has("poke") || t.has("sniper") || t.has("long-range")) counts.poke += 1;
      if (t.has("brawl") || t.has("melee")) counts.brawl += 1;
      if (t.has("bunker") || t.has("turret")) counts.bunker += 2;
      if (t.has("flyer")) counts.flying += 2;
      if (t.has("sniper")) counts.sniper += 2;
      if (t.has("flank")) counts.flank += 1;
    }
    if (heroes.some((h) => h.tags.includes("flyer")) && heroes.some((h) => h.tags.includes("flyer-synergy"))) {
      counts.flying += 3;
    }
    const ranked = Object.entries(counts).sort((a, b) => b[1] - a[1]);
    const primary = ranked[0][1] > 0 ? ranked[0][0] : "flex";
    const secondary = ranked[1] && ranked[1][1] > 0 ? ranked[1][0] : null;
    return { counts, primary, secondary, heroes };
  }

  const COMP_LABEL = {
    dive: { ja: "飛び込み", en: "Dive" },
    poke: { ja: "遠くから削る", en: "Poke" },
    brawl: { ja: "近距離の殴り合い", en: "Brawl" },
    bunker: { ja: "固まって守る", en: "Bunker" },
    flying: { ja: "空を飛ぶ", en: "Flying" },
    sniper: { ja: "スナイパー", en: "Sniper" },
    flank: { ja: "横から来る", en: "Flank" },
    flex: { ja: "ミックス", en: "Mix" },
  };

  const COMP_ANSWERS = {
    dive: {
      tank: ["orisa", "roadhog", "zarya", "dmon"],
      damage: ["cassidy", "mei", "torbjorn", "reaper"],
      support: ["brigitte", "ana", "moira", "wuyang"],
    },
    poke: {
      tank: ["winston", "dva", "doomfist", "wrecking-ball"],
      damage: ["genji", "tracer", "sombra", "venture"],
      support: ["lucio", "kiriko", "juno"],
    },
    brawl: {
      tank: ["mauga", "zarya", "junker-queen", "orisa"],
      damage: ["reaper", "mei", "pharah", "bastion"],
      support: ["ana", "moira", "lucio", "kiriko"],
    },
    bunker: {
      tank: ["wrecking-ball", "winston", "hazard", "dva"],
      damage: ["sombra", "venture", "junkrat", "pharah"],
      support: ["lucio", "kiriko", "sombra"],
    },
    flying: {
      tank: ["dva", "sigma", "domina"],
      damage: ["ashe", "widowmaker", "cassidy", "soldier-76", "sojourn"],
      support: ["baptiste", "mizuki", "ana", "illari"],
    },
    sniper: {
      tank: ["winston", "dva", "wrecking-ball", "doomfist"],
      damage: ["genji", "tracer", "sombra", "winston"],
      support: ["lucio", "kiriko", "moira"],
    },
    flank: {
      tank: ["orisa", "zarya", "roadhog"],
      damage: ["cassidy", "torbjorn", "mei", "symmetra"],
      support: ["brigitte", "moira", "baptiste"],
    },
  };

  function mapAffinity(hero, map, side) {
    if (!map) return { score: 0, reasons: [] };
    const traits = new Set(map.traits || []);
    const tags = new Set(hero.tags);
    let score = 0;
    const reasons = [];

    const bump = (cond, pts, ja) => {
      if (cond) {
        score += pts;
        if (pts >= 2) reasons.push(ja);
      }
    };

    bump(traits.has("longsight") && (tags.has("sniper") || tags.has("hitscan") || tags.has("long-range")), 4, `${map.nameJa}は長視線。${hero.nameJa}の射程が刺さる`);
    bump(traits.has("choke") && (tags.has("barrier") || tags.has("spam") || tags.has("freeze") || tags.has("brawl")), 3, `チョークマップで${hero.nameJa}のエリア制御が強い`);
    bump(traits.has("vertical") && (tags.has("flyer") || tags.has("mobile") || hero.key === "winston" || hero.key === "juno" || hero.key === "dva"), 3, `高低差のある地形で機動力が活きる`);
    bump(traits.has("environmental") && (tags.has("boop") || hero.key === "lucio" || hero.key === "junkrat" || hero.key === "hazard" || hero.key === "pharah"), 3, `崖・井戸があり環境キルが取りやすい`);
    bump(traits.has("close") && (tags.has("brawl") || tags.has("melee") || tags.has("tankbuster") || tags.has("beam")), 3, `近接寄りのマップで近距離キットが強い`);
    bump(traits.has("open") && tags.has("flyer"), 2, `開放空間でフライヤーが通りやすい`);
    bump(traits.has("open") && tags.has("sniper"), 3, `開放マップでスナイプが通りやすい`);
    bump(traits.has("flank") && (tags.has("flank") || tags.has("dive")), 2, `側面ルートが多くダイブ／フランカー向き`);
    bump(traits.has("highground") && (tags.has("hitscan") || tags.has("poke") || tags.has("high-ground") || tags.has("flyer")), 2, `高所を取れるヒーロー`);
    bump(traits.has("control") && tags.has("speed"), 2, `コントロールは回転力とスピードが重要`);
    bump(traits.has("flashpoint") && (tags.has("mobile") || tags.has("speed") || tags.has("flank")), 3, `フラッシュポイントはローテ性能が勝つ`);
    bump(traits.has("clash") && (tags.has("brawl") || tags.has("melee") || tags.has("barrier")), 3, `クラッシュは超近接ブロウルが強い`);
    bump(traits.has("push") && (tags.has("speed") || tags.has("brawl") || tags.has("mobile")), 2, `プッシュはボット周辺の近接とスピード`);

    bump(traits.has("longsight") && tags.has("melee") && !tags.has("mobile"), -3, `${map.nameJa}の長視線は近接に厳しい`);
    bump(traits.has("close") && tags.has("sniper"), -3, `近接マップでスナイパーは価値が落ちる`);
    bump(traits.has("choke") && tags.has("flyer"), -1, `閉鎖チョークでは空が取りにくい`);

    if (side === "attack" && (tags.has("dive") || tags.has("flank") || tags.has("speed"))) {
      score += 1;
    }
    if (side === "defend" && (tags.has("bunker") || tags.has("barrier") || tags.has("turret") || tags.has("sniper"))) {
      score += 1;
    }
    return { score, reasons };
  }

  function threatFrom(enemy, candidate) {
    const direct = explicitCounter(enemy.key, candidate.key);
    const tagged = tagCounter(enemy, candidate);
    return Math.max(direct, tagged.score * 0.7);
  }

  function powerAgainst(candidate, enemy) {
    const direct = explicitCounter(candidate.key, enemy.key);
    const tagged = tagCounter(candidate, enemy);
    const score = Math.max(direct, tagged.score);
    return { score, direct, rules: tagged.rules };
  }

  function recommend(state) {
    const myRole = state.myRole || "damage";
    const enemies = (state.enemies || []).map(heroByKey).filter(Boolean);
    const allies = (state.allies || []).map(heroByKey).filter(Boolean);
    const taken = new Set([...enemies.map((h) => h.key), ...allies.map((h) => h.key)]);
    const map = state.mapKey ? mapByKey(state.mapKey) : null;
    const side = state.side || "flex";
    const comp = detectComposition(enemies.map((h) => h.key));

    const answers = (COMP_ANSWERS[comp.primary] || {})[myRole] || [];

    const candidates = OW.heroes.filter((h) => h.role === myRole && !taken.has(h.key));

    const ranked = candidates.map((hero) => {
      let score = 24;
      const reasons = [];
      const vs = [];

      for (const enemy of enemies) {
        const hit = powerAgainst(hero, enemy);
        const threat = threatFrom(enemy, hero);
        const delta = hit.score * 4.6 - threat * 3.4;
        score += delta;
        vs.push({
          enemy: enemy.key,
          power: hit.score,
          threat,
          rules: hit.rules,
        });
        if (hit.score >= 4) {
          reasons.push(`${enemy.nameJa}に強い`);
        } else if (hit.score >= 3) {
          reasons.push(`${enemy.nameJa}との相性がいい`);
        }
        if (threat >= 4) {
          reasons.push(`${enemy.nameJa}には注意`);
          score -= 2;
        }
      }

      if (answers.includes(hero.key)) {
        score += 6;
        reasons.push("この敵の戦い方に向いている");
      }

      const mapFit = mapAffinity(hero, map, side);
      score += mapFit.score * 1.6;
      reasons.push(...mapFit.reasons.slice(0, 2));

      if (comp.primary === "flying" && hero.tags.includes("hitscan")) {
        score += 5;
        reasons.push("空を飛ぶ相手には、弾がまっすぐ飛ぶヒーロー");
      }
      if (comp.counts.flying >= 2 && hero.tags.includes("hitscan")) {
        score += 6;
        reasons.push("空の2人には、弾がまっすぐ飛ぶヒーロー");
      }
      if (comp.primary === "dive" && (hero.tags.includes("anti-dive") || hero.tags.includes("cc") || hero.tags.includes("sleeper") || hero.tags.includes("anti-flank"))) {
        score += 4;
      }
      if (comp.primary === "sniper" && (hero.tags.includes("dive") || hero.tags.includes("flank"))) {
        score += 4;
        reasons.push("スナイパーの後ろに回れる");
      }

      const uniq = [];
      const seen = new Set();
      for (const r of reasons) {
        if (!seen.has(r)) {
          seen.add(r);
          uniq.push(r);
        }
      }

      return {
        hero,
        raw: score,
        reasons: uniq.slice(0, 4),
        vs,
        mapFit: mapFit.score,
      };
    });

    ranked.sort((a, b) => b.raw - a.raw);
    const topRaw = ranked[0] ? ranked[0].raw : 1;
    const floor = ranked.length ? ranked[Math.min(ranked.length - 1, 12)].raw : 0;
    for (const row of ranked) {
      const t = (row.raw - floor) / Math.max(8, topRaw - floor);
      row.score = Math.round(Math.max(12, Math.min(99, 28 + t * 71)));
    }

    const weaknesses = [];
    if (comp.primary === "flying") weaknesses.push({ ja: "空を飛んでくる相手。下だけ見ていると負ける。", en: "Without hitscan, flyers own the sky" });
    if (comp.primary === "dive") weaknesses.push({ ja: "一気に後ろへ飛び込んでくる。回復役の横で待つ。", en: "No peel means the backline dies" });
    if (comp.primary === "brawl") weaknesses.push({ ja: "真正面の殴り合い。回復を止められると崩れる。", en: "Anti-heal and high ground break brawl" });
    if (comp.primary === "poke") weaknesses.push({ ja: "遠くから削ってくる。近づけると弱い。", en: "Poke loses if you close the gap" });
    if (comp.primary === "bunker") weaknesses.push({ ja: "固まって守りを固める。横や上から崩す。", en: "Hack, burrow, and air collapse bunker" });
    if (comp.primary === "sniper") weaknesses.push({ ja: "遠くから頭を狙う。箱の陰を伝って近づく。", en: "Cover and dive mute snipers" });

    return {
      comp,
      compLabel: COMP_LABEL[comp.primary] || COMP_LABEL.flex,
      weaknesses,
      picks: ranked.slice(0, 5),
      avoid: ranked.slice(-3).reverse(),
      map,
    };
  }

  function analyzeScreenshotText(text) {
    if (!text) return { mapKey: null, heroKeys: [] };
    const fold = (s) =>
      String(s || "")
        .toLowerCase()
        .replace(/[’']/g, "'")
        .replace(/[：]/g, ":")
        .replace(/[　]/g, " ");
    const lower = fold(text);
    let mapKey = null;
    let bestLen = 0;
    for (const m of OW.maps) {
      const extra = m.key === "route-66" ? ["route 66", "route66", "ルート66"] : [];
      for (const n of [m.name, m.nameJa, m.key.replace(/-/g, " "), m.key.replace(/-/g, ""), ...extra]) {
        const needle = fold(n).replace(/:/g, "");
        if (needle.length >= 4 && lower.replace(/:/g, "").includes(needle) && needle.length > bestLen) {
          mapKey = m.key;
          bestLen = needle.length;
        }
      }
    }
    const heroKeys = [];
    const names = [];
    for (const h of OW.heroes) {
      names.push([h.key, h.name.toLowerCase()]);
      names.push([h.key, h.nameJa.toLowerCase()]);
      names.push([h.key, h.key.replace(/-/g, " ")]);
      names.push([h.key, h.name.toLowerCase().replace(/[:']/g, "")]);
    }
    names.sort((a, b) => b[1].length - a[1].length);
    const used = new Set();
    for (const [key, name] of names) {
      if (used.has(key)) continue;
      if (name.length < 3) continue;
      if (lower.includes(name)) {
        heroKeys.push(key);
        used.add(key);
      }
    }
    return { mapKey, heroKeys };
  }

  window.Engine = {
    heroByKey,
    mapByKey,
    detectComposition,
    recommend,
    analyzeScreenshotText,
    COMP_LABEL,
    ROLE_ORDER,
  };
})();
